File: Data Structures & Algorithms/subsets/submission-18.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums) { 
        const part = [] 
        const res = [] 
        this.dfs(0, nums, part, res) 
        return res
    } 

    dfs(start, nums, part, res) {
        res.push([...part]) 

        for (let i = start; i < nums.length; i ++) {
            part.push(nums[i]) 
            this.dfs(i + 1, nums, part, res) 
            part.pop()
        }
    }
}

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-18.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        memo = {}
        def dfs(prev, i): 
            if (prev, i) in memo: 
                return memo[(prev, i)] 
            if i == len(nums): 
                return 0 

            res = dfs(prev, i + 1) 

            if prev == -1 or nums[prev] < nums[i]: 
                res = max(res, 1 + dfs(i, i + 1)) 
            memo[(prev, i)] = res
            return res 

        return dfs(-1, 0)